contracts/asset-proxy/contracts/src/libs/LibAssetData.sol

```diff
@@ -0,0 +1,420 @@
+import "@0x/contracts-erc721/contracts/src/interfaces/IERC721Token.sol";
+
+
+library LibAssetData {
+    bytes4 constant public ERC20_PROXY_ID = bytes4(keccak256("ERC20Token(address)"));
+    bytes4 constant public ERC721_PROXY_ID = bytes4(keccak256("ERC721Token(address,uint256)"));
+    bytes4 constant public ERC1155_PROXY_ID = bytes4(keccak256("ERC1155Assets(address,uint256[],uint256[],bytes)"));
+    bytes4 constant public MULTI_ASSET_PROXY_ID = bytes4(keccak256("MultiAsset(uint256[],bytes[])"));
+
+    /// @dev Returns the owner's balance of the token(s) specified in
+    ///     assetData.  When the asset data contains multiple tokens (eg in
+    ///     ERC1155 or Multi-Asset), the return value indicates how many
+    ///     complete "baskets" of those tokens are owned by owner.
+    /// @param owner Owner of the tokens specified by assetData.
+    /// @param assetData Description of tokens, per the AssetProxy contract
+    ///     specification.
+    /// @return Number of tokens (or token baskets) held by owner.
+    function getBalance(address owner, bytes memory assetData)
+        public
+        view
+        returns (uint256 balance)
+    {
+        bytes4 proxyId = LibBytes.readBytes4(assetData, 0);
+        if (proxyId == ERC20_PROXY_ID) {
+            address tokenAddress = LibBytes.readAddress(assetData, 16);
+            return IERC20Token(tokenAddress).balanceOf(owner);
+        } else if (proxyId == ERC721_PROXY_ID) {
+            (, address tokenAddress, uint256 tokenId) = decodeERC721AssetData(assetData);
+            return getERC721TokenOwner(tokenAddress, tokenId) == owner ? 1 : 0;
+        } else if (proxyId == ERC1155_PROXY_ID) {
+            uint256 lowestTokenBalance = 0;
+            (
+                ,
+                address tokenAddress,
+                uint256[] memory tokenIds,
+                uint256[] memory tokenValues,
+            ) = decodeERC1155AssetData(assetData);
+            for (uint256 i = 0; i < tokenIds.length; i++) {
+                uint256 tokenBalance = IERC1155(tokenAddress).balanceOf(owner, tokenIds[i]) / tokenValues[i];
+                if (tokenBalance < lowestTokenBalance || lowestTokenBalance == 0) {
+                    lowestTokenBalance = tokenBalance;
+                }
+            }
+            return lowestTokenBalance;
+        } else if (proxyId == MULTI_ASSET_PROXY_ID) {
+            uint256 lowestAssetBalance = 0;
+            (, uint256[] memory assetAmounts, bytes[] memory nestedAssetData) = decodeMultiAssetData(assetData);
+            for (uint256 i = 0; i < nestedAssetData.length; i++) {
+                uint256 assetBalance = getBalance(owner, nestedAssetData[i]) / assetAmounts[i];
+                if (assetBalance < lowestAssetBalance || lowestAssetBalance == 0) {
+                    lowestAssetBalance = assetBalance;
+                }
+            }
+            return lowestAssetBalance;
+        } else {
+            revert("Unsupported proxy identifier");
```

Comment:
Sorry, missed this first round. Our convention is to revert with `"ALL_CAPS_SNAKE_CASE"` error messages.